In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

<h1> Wczytanie danych <h1>

In [2]:
data=pd.read_csv("data.csv")
data.head(5)

Unnamed: 0  pclass  survived  \
0           0       1         1   
1           1       1         1   
2           2       1         0   
3           3       1         0   
4           4       1         0   

                                              name     sex      age  sibsp  \
0                    Allen, Miss. Elisabeth Walton  female  29.0000      0   
1                   Allison, Master. Hudson Trevor    male   0.9167      1   
2                     Allison, Miss. Helen Loraine  female   2.0000      1   
3             Allison, Mr. Hudson Joshua Creighton    male  30.0000      1   
4  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000      1   

   parch  ticket      fare    cabin embarked boat   body  \
0      0   24160  211.3375       B5        S    2    NaN   
1      2  113781  151.5500  C22 C26        S   11    NaN   
2      2  113781  151.5500  C22 C26        S  NaN    NaN   
3      2  113781  151.5500  C22 C26        S  NaN  135.0   
4      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  KolumnNull CabinReduced  
0                     St Louis, MO           0            B  
1  Montreal, PQ / Chesterville, ON           0            C  
2  Montreal, PQ / Chesterville, ON           1            C  
3  Montreal, PQ / Chesterville, ON           1            C  
4  Montreal, PQ / Chesterville, ON           1            C

Funkcja train_test_split przyjmuje parametry:
- test_size
- train_size (oba określają rozmiar danych testowych/treningowych w stosunku do cąłości zbioru, jeżeli jeden tylko zostanie ukreślony, drugi jest dopełnieniem do całości; wartość z przedziału 0-1)
- arrays - zbiór danych (dataframe, numpy array, lista itd)
- random_state - int odpowiadający za poziom przetasowania danych przed dokonaniem podziału na zbiory

Wartość zwracana to dwa obiekty - zbiór testowy i zbiór treningowy

In [3]:
col_name = ['sex','cabin','CabinReduced']
data[col_name]

sex    cabin CabinReduced
0     female       B5            B
1       male  C22 C26            C
2     female  C22 C26            C
3       male  C22 C26            C
4     female  C22 C26            C
...      ...      ...          ...
1304  female      NaN            n
1305  female      NaN            n
1306    male      NaN            n
1307    male      NaN            n
1308    male      NaN            n

[1309 rows x 3 columns]

<h1> Podział zbioru na zbiór treningowy i zbiór testowy <h1>

In [4]:
X_train, X_test, y_train, y_test = train_test_split(data[col_name],data['survived'],test_size=0.30,random_state = 0)

In [5]:
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)


(916, 3) (393, 3) (916,) (393,)


X_test i y_test liczą po 393 obserwacji, co stanowi 30% całego zbioru danych (co zgadza się z tym co zostało podane w poleceniu funkcji train_test_split)


<h1>Kardynalność zmiennych<h1>

In [6]:
Unique_test = [x for x in X_test.sex.unique() if x not in X_train.sex.unique()]
len(Unique_test)

0

In [7]:
Unique_test = [x for x in X_test.cabin.unique() if x not in X_train.cabin.unique()]
len(Unique_test)

37

In [8]:
Unique_test = [x for x in X_test.CabinReduced.unique() if x not in X_train.CabinReduced.unique()]
len(Unique_test)

0

Rozkład cech w zależności od kardynalności dla zmiennej sex i cabin jest równomierny, natomiast dla CabinReduced nie

<h1> Kodowanie zmiennych kategorycznych do zmiennych liczbowych <h1>

In [10]:
Dict_cabin_test={}
for i,x in enumerate(X_test['cabin'].unique()):
    Dict_cabin_test[x]=i
Dict_cabin_test

{nan: 0,
 'G6': 1,
 'E12': 2,
 'C104': 3,
 'B57 B59 B63 B66': 4,
 'A31': 5,
 'B96 B98': 6,
 'B18': 7,
 'F33': 8,
 'C124': 9,
 'D11': 10,
 'D48': 11,
 'D10 D12': 12,
 'D30': 13,
 'B38': 14,
 'D45': 15,
 'D15': 16,
 'C50': 17,
 'C31': 18,
 'E24': 19,
 'B82 B84': 20,
 'C85': 21,
 'C22 C26': 22,
 'A32': 23,
 'C53': 24,
 'B78': 25,
 'B10': 26,
 'C70': 27,
 'A23': 28,
 'B41': 29,
 'C106': 30,
 'D37': 31,
 'C46': 32,
 'B20': 33,
 'E58': 34,
 'D': 35,
 'B11': 36,
 'C101': 37,
 'B49': 38,
 'E25': 39,
 'F E69': 40,
 'B80': 41,
 'C80': 42,
 'C52': 43,
 'C2': 44,
 'E39 E41': 45,
 'F4': 46,
 'D22': 47,
 'B51 B53 B55': 48,
 'E40': 49,
 'F G73': 50,
 'A19': 51,
 'C32': 52,
 'B79': 53,
 'C62 C64': 54,
 'E8': 55,
 'D28': 56,
 'C45': 57,
 'B22': 58,
 'C23 C25 C27': 59,
 'E44': 60,
 'B45': 61,
 'B39': 62,
 'B58 B60': 63,
 'F2': 64,
 'C47': 65,
 'B101': 66,
 'D21': 67,
 'A7': 68,
 'E52': 69,
 'F38': 70}

In [11]:
Dict_cabin_reduced_test={}
for i,x in enumerate(X_test['CabinReduced'].unique()):
    Dict_cabin_reduced_test[x]=i
Dict_cabin_reduced_test

{'n': 0, 'G': 1, 'E': 2, 'C': 3, 'B': 4, 'A': 5, 'F': 6, 'D': 7}

In [12]:
Dict_cabin_reduced_train={}
for i,x in enumerate(X_train['CabinReduced'].unique()):
    Dict_cabin_reduced_train[x]=i
Dict_cabin_reduced_train

{'n': 0, 'E': 1, 'C': 2, 'D': 3, 'B': 4, 'A': 5, 'F': 6, 'T': 7, 'G': 8}

In [13]:
Dict_cabin_train={}
for i,x in enumerate(X_train['cabin'].unique()):
    Dict_cabin_train[x]=i
Dict_cabin_train

{nan: 0,
 'E36': 1,
 'C68': 2,
 'E24': 3,
 'C22 C26': 4,
 'D38': 5,
 'B50': 6,
 'A24': 7,
 'C111': 8,
 'F': 9,
 'C6': 10,
 'C87': 11,
 'E8': 12,
 'B45': 13,
 'C93': 14,
 'D28': 15,
 'D36': 16,
 'C125': 17,
 'B35': 18,
 'T': 19,
 'B73': 20,
 'B57 B59 B63 B66': 21,
 'A26': 22,
 'A18': 23,
 'B96 B98': 24,
 'G6': 25,
 'C78': 26,
 'C101': 27,
 'D9': 28,
 'D33': 29,
 'C128': 30,
 'E50': 31,
 'B26': 32,
 'B69': 33,
 'E121': 34,
 'C123': 35,
 'B94': 36,
 'A34': 37,
 'D': 38,
 'C39': 39,
 'D43': 40,
 'E31': 41,
 'B5': 42,
 'D17': 43,
 'F33': 44,
 'E44': 45,
 'D7': 46,
 'A21': 47,
 'D34': 48,
 'A29': 49,
 'D35': 50,
 'A11': 51,
 'B51 B53 B55': 52,
 'D46': 53,
 'E60': 54,
 'C30': 55,
 'D26': 56,
 'E68': 57,
 'A9': 58,
 'B71': 59,
 'D37': 60,
 'F2': 61,
 'C55 C57': 62,
 'C89': 63,
 'C124': 64,
 'C23 C25 C27': 65,
 'C126': 66,
 'E49': 67,
 'F E46': 68,
 'E46': 69,
 'D19': 70,
 'B58 B60': 71,
 'C82': 72,
 'B52 B54 B56': 73,
 'C92': 74,
 'E45': 75,
 'F G73': 76,
 'C65': 77,
 'E25': 78,
 'B3': 79,
 'D

In [14]:
Dict_sex_test={}
for i, x in enumerate(X_test.sex.unique()):
    Dict_sex_test[x]=i
Dict_sex_test

{'male': 0, 'female': 1}

In [15]:
Dict_sex_train={}
for i, x in enumerate(X_train.sex.unique()):
    Dict_sex_train[x]=i
Dict_sex_train

{'female': 0, 'male': 1}

<h1> Mapowanie zmienych <h1>

In [16]:
X_test['cabin_map']=X_test['cabin'].map(Dict_cabin_test)
X_train['cabin_map']=X_train['cabin'].map(Dict_cabin_train)
X_train['cabin_map_reduced']=X_train['CabinReduced'].map(Dict_cabin_reduced_train)
X_test['cabin_map_reduced']=X_test['CabinReduced'].map(Dict_cabin_reduced_test)
X_test['sex_map']=X_test['sex'].map(Dict_sex_test)
X_train['sex_map']=X_train['sex'].map(Dict_sex_train)

In [17]:
print(f'Nulle dla zbioru testowego :\n{pd.isnull(X_test).sum()}\n')
print(f'Nulle dla zbioru treningowego :\n{pd.isnull(X_train).sum()}')


Nulle dla zbioru testowego :
sex                    0
cabin                312
CabinReduced           0
cabin_map              0
cabin_map_reduced      0
sex_map                0
dtype: int64

Nulle dla zbioru treningowego :
sex                    0
cabin                702
CabinReduced           0
cabin_map              0
cabin_map_reduced      0
sex_map                0
dtype: int64


<p> W moim przypadku brakujące wartości dla zmiennych powstałych z mapowania już są ustawione jako wartość zero, dlatego w powyższym podsumowaniu nie ma wartości NaN dla nowych zmiennych <p>

<h1> Podsumowanie <h1>

In [18]:
print('Liczba etykiet zmiennej CabinReduced test: {}'.format(len(X_test["CabinReduced"].unique())))
print('Liczba etykiet zmiennej CabinReduced train: {}'.format(len(X_train["CabinReduced"].unique())))
print('Liczba etykiet zmiennej cabin_map_reduced test: {}'.format(len(X_test["cabin_map_reduced"].unique())))
print('Liczba etykiet zmiennej cabin_map_reduced train: {}'.format(len(X_train["cabin_map_reduced"].unique())))


Liczba etykiet zmiennej CabinReduced test: 8
Liczba etykiet zmiennej CabinReduced train: 9
Liczba etykiet zmiennej cabin_map_reduced test: 8
Liczba etykiet zmiennej cabin_map_reduced train: 9


In [19]:
print('Liczba etykiet zmiennej cabin test: {}'.format(len(X_test["cabin"].unique())))
print('Liczba etykiet zmiennej cabin train: {}'.format(len(X_train["cabin"].unique())))
print('Liczba etykiet zmiennej cabin_map test: {}'.format(len(X_test["cabin_map"].unique())))
print('Liczba etykiet zmiennej cabin_map train: {}'.format(len(X_train["cabin_map"].unique())))


Liczba etykiet zmiennej cabin test: 71
Liczba etykiet zmiennej cabin train: 151
Liczba etykiet zmiennej cabin_map test: 71
Liczba etykiet zmiennej cabin_map train: 151


In [20]:
print('Liczba etykiet zmiennej sex test: {}'.format(len(X_test["sex"].unique())))
print('Liczba etykiet zmiennej sex train: {}'.format(len(X_train["sex"].unique())))
print('Liczba etykiet zmiennej sex_map test: {}'.format(len(X_test["sex_map"].unique())))
print('Liczba etykiet zmiennej sex_map train: {}'.format(len(X_train["sex_map"].unique())))


Liczba etykiet zmiennej sex test: 2
Liczba etykiet zmiennej sex train: 2
Liczba etykiet zmiennej sex_map test: 2
Liczba etykiet zmiennej sex_map train: 2


<p> Ogólnie zbiór treningowy ma większą liczebność niż zbiór testowy, ponieważ jest znacznie większy (stosunek 30% do 70%).  Liczebność etykiet zmiennej cabin uległa znacznej redukcji. Z 151 do 8 dla zbioru treningowego i z 71 do 8 dla zbioru testowego. Proces redukcji etykiet i zamiany zmiennych kategorycznych na zmienne liczbowe ma duże znaczenie dla tworzenia modelu i predykcji. Zmniejszenie liczebności etykiet chroni przed overfittingiem i dominacją zmiennej o dużej kardynalności nad zmiennymi o mniejszej kardynalności. Natomiast zmiana wartości na liczbowe upraszcza tworzenie modelu.<p>